### PR TITLE
Feature/gin server

### PR DESCRIPTION
**Pull request by ymichaeli**

### CODE REVIEW

File: cmd/server/main.go [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"OnboardingExercise/pkg/api"
+	"github.com/pkg/errors"
+)
+
+func main() {
+	server := api.NewServer()
+
+	if err := server.Start(); err != nil {
+		panic(errors.WithStack(errors.Wrap(err, "Couldn't start server")))
```
Comment (shai-ltx) [inline cmd/server/main.go]: with stack and wrap returns the same struct

File: pkg/api/lifecycle/handler.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package lifecycle
+
+import (
+	"OnboardingExercise/pkg/service/lifecycle"
+	"github.com/gin-gonic/gin"
+	"net/http"
+)
+
+type Handler struct {
```
Comment (shai-ltx) [inline pkg/api/lifecycle/handler.go:10]: comments

File: pkg/api/lifecycle/handler.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	}
+	c.JSON(http.StatusServiceUnavailable, gin.H{"error": "Service unavailable"})
+}
+
+func (h Handler) readiness(c *gin.Context) {
+	if h.service.IsReady() {
+		c.JSON(http.StatusOK, gin.H{"status": "ok"})
+		return
+	}
+	c.JSON(http.StatusServiceUnavailable, gin.H{"error": "Service unavailable"})
```
Comment (shai-ltx) [inline pkg/api/lifecycle/handler.go]: json in response is not needed if it doesn't add more info than the http code

File: pkg/api/middlewares/error_handler.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"OnboardingExercise/pkg/custom_errors"
+	"fmt"
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+	"net/http"
+)
+
+func ErrorHandler(c *gin.Context) {
```
Comment (shai-ltx) [inline pkg/api/middlewares/error_handler.go]: comments

File: pkg/api/middlewares/error_handler.go [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+	c.Next()
+
+	if len(c.Errors) > 0 {
+		err := c.Errors[0].Err // to simplify I assume there is only one Error
+
+		// log the full error with stacktrace
+		fmt.Printf("%+v\n", errors.WithStack(err))
+
+		if errors.As(err, &custom_errors.NotFoundError{}) {
+			c.JSON(http.StatusNotFound, gin.H{"error": err.Error()})
```
Comment (shai-ltx) [inline pkg/api/middlewares/error_handler.go]: err.Error might expose internal data that you don't want to share with the clients (e.g. stack trace)

File: pkg/api/middlewares/error_handler_test.go [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+)
+
+type MiddlewareTestSuite struct {
```
Comment (shai-ltx) [inline pkg/api/middlewares/error_handler_test.go]: why do you need a suite here? 
each test is an independent  unit test

File: pkg/api/profile/handler.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+func (handler *Handler) GetAllProfiles(c *gin.Context) {
+	profiles := handler.service.GetAllProfiles()
+
+	c.JSON(http.StatusOK, profiles)
+}
+
+func (handler *Handler) GetProfileByUserID(c *gin.Context) {
+	userId := c.Param("userId")
```
Comment (shai-ltx) [inline pkg/api/profile/handler.go]: use binding

File: pkg/api/profile/handler.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+
+	c.JSON(http.StatusOK, userProfile)
+}
+
+func (handler *Handler) CreateProfile(c *gin.Context) {
+	var newProfile profile_repository.Profile
```
Comment (shai-ltx) [inline pkg/api/profile/handler.go]: this layer should not be aware of db models

File: pkg/api/profile/handler.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+		return
+	}
+
+	newProfile = handler.service.CreateProfile(newProfile)
+	c.JSON(http.StatusCreated, newProfile)
+
+}
+
+func (handler *Handler) UpdateProfile(c *gin.Context) {
+	var newProfile profile_repository.Profile
```
Comment (shai-ltx) [inline pkg/api/profile/handler.go]: same comment

File: pkg/api/server.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	server := Server{engine}
+
+	server.initMiddlewares()
+	server.initRoutes()
+
+	return server
+}
+
+func (server Server) Start() error {
+	return server.engine.Run("localhost:8080")
```
Comment (shai-ltx) [inline pkg/api/server.go]: inject this value to the server from main

File: pkg/repository/profile/model.go [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package profile_repository
+
+// Profile struct contains UserId which is not required because we generate it when the client wants to create a new user
+// is it ok, or I should create a partial Profile struct without UserId?
+type Profile struct {
```
Comment (shai-ltx) [inline pkg/repository/profile/model.go:3]: repo model should not have binding

File: pkg/repository/profile/repository.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+// Repository Note: I thought about another solution for the address changed case
+// in a regular case the Repository won't contain something that should change
+// it will contain a constant db client
+// so maybe a better solution will be to extract the profiles slice from the repository and make it global?
+type Repository struct {
+	profiles []Profile // will be changed to db client
+}
+
+func NewDAL(profiles []Profile) Repository {
```
Comment (shai-ltx) [inline pkg/repository/profile/repository.go]: NewRepository

File: pkg/repository/profile/repository.go [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+func (repository *Repository) GetProfileByUserID(userId string) (Profile, error) {
+	for _, profile := range repository.GetAllProfiles() {
+		if profile.UserId == userId {
+			return profile, nil
+		}
+	}
+
+	return Profile{}, custom_errors.NotFoundError{Err: fmt.Sprintf("user %s does not exist", userId)}
```
Comment (shai-ltx) [inline pkg/repository/profile/repository.go]: I'd use errors.Wrap so we will have a stack trace

**Review by shai-ltx — CHANGES_REQUESTED** [review]

File: cmd/server/main.go (hunk 1 above)
Comment (ymichaeli) [reply cmd/server/main.go]: done

File: pkg/api/lifecycle/handler.go (hunk 2 above)
Comment (ymichaeli) [reply pkg/api/lifecycle/handler.go:10]: done

File: pkg/api/lifecycle/handler.go (hunk 3 above)
Comment (ymichaeli) [reply pkg/api/lifecycle/handler.go]: removed

File: pkg/api/middlewares/error_handler.go (hunk 4 above)
Comment (ymichaeli) [reply pkg/api/middlewares/error_handler.go]: done

File: pkg/api/middlewares/error_handler.go (hunk 5 above)
Comment (ymichaeli) [reply pkg/api/middlewares/error_handler.go]: done

File: pkg/api/middlewares/error_handler_test.go (hunk 6 above)
Comment (ymichaeli) [reply pkg/api/middlewares/error_handler_test.go]: removed

File: pkg/api/profile/handler.go (hunk 8 above)
Comment (ymichaeli) [reply pkg/api/profile/handler.go]: done

File: pkg/api/profile/handler.go (hunk 9 above)
Comment (ymichaeli) [reply pkg/api/profile/handler.go]: done

File: pkg/api/server.go (hunk 10 above)
Comment (ymichaeli) [reply pkg/api/server.go]: done

File: pkg/repository/profile/repository.go (hunk 13 above)
Comment (ymichaeli) [reply pkg/repository/profile/repository.go]: I added the stack trace to the error initlaization

File: pkg/custom_errors/bad_request_error.go [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+func (e BadRequestError) Error() string {
+	if e.err != "" {
+		return e.err
+	}
+	return e.publicMessage
+}
+
+func (e BadRequestError) DisplayMessage() string {
```
Comment (shai-ltx) [inline pkg/custom_errors/bad_request_error.go:22]: comments

File: pkg/repository/profile/model.go (hunk 11 above)
Comment (shai-ltx) [reply pkg/repository/profile/model.go:3]: for next phase consider using https://pkg.go.dev/github.com/kisielk/sqlstruct (although we currently don't use it in our code)

**Review by shai-ltx — APPROVED** [review]

**Pull request merged**